---
title: "Cancer Personalized Drug Recommendation (CPDR)"
author: "Ruzhen Chen"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %. VignetteIndexEntry{CPDR}
  %. VignetteEngine{knitr::rmarkdown}
  %. VignetteEncoding{UTF-8}
---

### Ruzhen Chen 

# Package overview
Due to cancer heterogeneity, only some patients can benefit from drug therapy. The personalized drug use is important for improving the treatment response rate of cancer patients. Patient transcriptome has recently demonstrated its value in guiding personalized drug use, and Connectivity map (CMAP) is a reliable computational approach for drug recommendation. However, there is still no personalized drug recommendation tool based on patient transcriptomic profiles and CMAP. To fill this gap, here we proposed such a feasible workflow and a user-friendly R package - Cancer Personalized Drug Recommendation (CPDR). CPDR has three features. 1) It identifies the individual disease signature by using the patient subgroup with transcriptomic profiles similar to that of the input patient. 2) Transcriptomic profile purification is supported for the subgroup with high infiltration of non-cancerous cells.  3) It supports in silico drug efficacy assessment using drug sensitivity data of cancer cell lines.

# Workflow 
We demonstrated the pipeline of CPDR with the aid of a dataset from GEO (GSE164541), containing gene expression profiles of 5 patients with colorectal cancer.

## 1. Data download
### 1.1 Clinical patients' RNA-seq count data (GSE164541)
```{r eval=FALSE,message=FALSE,warning=FALSE}

library(CPDR)
library(GEOquery)
library(org.Hs.eg.db)

getGEOSuppFiles('GSE164541', makeDirectory = TRUE, fetch_files = TRUE)
clinical <- read.csv("./GSE164541/GSE164541_ANT_count.csv.gz")
clinical <- inner_join(clinical, 
                       bitr(clinical$ENSEMBL, 
                            fromType = "ENSEMBL",
                            toType = "SYMBOL", 
                            OrgDb = org.Hs.eg.db),by = "ENSEMBL")
clinical <- clinical[!duplicated(clinical$SYMBOL),]
row.names(clinical) <- clinical$SYMBOL
clinical_profile_set <- clinical[,grep("PT",colnames(clinical))]

```


### 1.2 Background data
``` {r eval=FALSE,message=FALSE,warning=FALSE}

download_db(pset = c('CCLE','PRISM'), # pharmacogenetic data
            tset = 'coadread',        # TCGA data. View(CPDR::TCGA_sets) for available cancer type
            nset = 'GTEX',            # GTEX data
            saveDir = '.',
            verbose = T)

# read downloaded TCGA dataset
exdir <- cBioPortalData::untarStudy("./CPDR_db/TCGA/25dc53d4417_coadread_tcga_pan_can_atlas_2018.tar.gz")
coadread <- cBioPortalData::loadStudy(exdir)

```


## 2. Identification of robust individual disease signals
### 2.1 Preprocessing
``` {r eval=FALSE,message=FALSE,warning=FALSE}

pmat = select_db(Assay = coadread, 
                 cmat = clinical_profile_set,
                 removeBatchEffect = TRUE, 
                 OrgDb = org.Hs.eg.db,
                 minSampleSize = 10,
                 MSI_status = NULL,
                 driver_gene = NULL,
                 MUT_status = NULL,
                 CNA_status = NULL)
gc()

```

### 2.2 NMF subtyping
``` {r eval=FALSE,message=FALSE,warning=FALSE,fig.show = "hold"}

result = get_NMF(mat = pmat$mat,
                 method = 'MAD',
                 clusterNum = 4, 
                 seed = 3211232,
                 nrun = 10,
                 doPlot = T)

```

### 2.3 Identification of subgroups
``` {r eval=FALSE,message=FALSE,warning=FALSE,fig.show = "hold"}

subgroup = get_subgroup(cmat = pmat$cmat, 
                        mat = pmat$mat, 
                        subtype = result, 
                        k = 10,
                        biopsy = "COLON",
                        adjacent = F,
                        doplot = T,
                        db.path = '.',
                        OrgDB = org.Hs.eg.db)


```

### 2.4 Estimation of non-cancerous infiltration
``` {r eval=FALSE,message=FALSE,warning=FALSE,fig.show = "hold"}

es = get_estimateScore(mat = pmat$mat, subtype=result, doplot = T)

```

### 2.5 Purification of subgroups 
``` {r eval=FALSE,message=FALSE,warning=FALSE,results = "asis"}
# subgroup simple version
subgroup_simple = lapply(subgroup, function(x){
  lapply(x, function(y){
    gene = intersect(row.names(y),row.names(octad.db::lincs_signatures))
    return(y[gene,])
  })
})

head(subgroup_simple[[1]]$case)

```

``` {r eval=FALSE,message=TRUE,warning=TRUE}
purify_data = get_puretumor(subgroup = subgroup_simple)

```

### 2.6 Differential expression analysis
``` {r eval=FALSE,message=FALSE,warning=FALSE}
signature = get_diff(data = purify_data, 
                     DE_method = 'limma', 
                     normalize_samples = TRUE, 
                     threshold_log2foldchange = 2, 
                     threshold_pval = 1, 
                     threshold_adjpval = 0.05)

```

## 3. Screening of candidate agents by reversing signals
``` {r eval=FALSE,message=FALSE,warning=FALSE}
sRGES = lapply(signature, function(x, LINCS){
  return(get_reverse_score(dz_signature = x,
                           max_gene_size=500,
                           permutations = 10000,
                           LINCS_data = LINCS)) 
  },LINCS = NULL)
gc()

```

## 4. Assessment of drug efficacy
### 4.1 Calculation of inidividual-related cell lines
``` {r eval=FALSE,message=FALSE,warning=FALSE}
cell_info = get_cell(cmat = pmat$cmat,
                     db.path = './CPDR_db/Pharmacogenomic',
                     removeBatchEffect = FALSE,
                     orgDB = org.Hs.eg.db)

```

### 4.2 Evaluattion of drug effectiveness at cell-line level
``` {r eval=FALSE,message=FALSE,warning=FALSE,fig.show = "hold",fig.height = 4, fig.width = 4}
cor = drugcorTest(mysRGES=sRGES, topline = topline, cell_info = cell_info)
draw_cor_map(sRGES[[1]],topline[1], cell_info = cell_info)

```

# Session information
Here is the output of sessionInfo on the system where this document was compiled:
```{r eval=TRUE} 
sessionInfo()
```
